### DISCUSSION TITLE
Inconsistent thermal divergence in low-Mach formulation

### DISCUSSION THREAD

**Discussion by bnkeeton** [Nek5000/nekRS#586] [General]
I've been using the low-Mach formulation in nekRS v23 and found that the evaluation of the thermal divergence seems incorrect for the problem I'm working on. To test this, I've set up a very simple case in a cylindrical domain with a stagnant initial flow superimposed by a spherical hot spot. I run the case for a single small time step, and then output the thermal divergence and compare it to the identical case run with nek5000 by plotting along the line [0,0,0] to [6,0,0]. I have confirmed through separate software that the nek5000 solution is correct. A plot showing the discrepancy is attached, along with both sets of case files. The red curve represents the correct, symmetrical nek5000 thermal divergence, and the blue is nekRS.

One additional item I found interesting is that if you set the thermal conductivity `k1` to be constant by changing `sigma1=0` in both cases, the two codes then produce identical values for the divergence. I tried inspecting the relevant kernel `qtlHex3D` but I do not see anything off.

<img width="450" alt="div" src="https://github.com/user-attachments/assets/d24ada10-100f-4658-b676-71d70455a3d4">

[nek5000_jet_1.zip](https://github.com/user-attachments/files/17048732/nek5000_jet_1.zip)
[nekrs_jet_1.zip](https://github.com/user-attachments/files/17048733/nekrs_jet_1.zip)

Thanks,
Ben Keeton

**Comment by stgeke** [comment]
Assuming the difference in qtl appears after the first timestep: Does the temperature match?

**Reply by bnkeeton** [reply]
Hi Stefan, yes, the temperature and thermal conductivity after the first time step are identical, and are shown below in first and second plots, respectively.

Given that $Q_T=\frac{1}{T} \frac{DT}{Dt}= \frac{1}{\rho T} \nabla \cdot (k \nabla T)$ for the single-component mixture considered in this problem, we should expect the thermal divergence to be identical.

<img width="400" alt="temp" src="https://github.com/user-attachments/assets/f3b8b894-a84d-47f7-afbd-01d896e66293">
<img width="400" alt="cnd" src="https://github.com/user-attachments/assets/9111cd76-c65e-4c4d-8241-bd0dcce1749f">

Thanks,
Ben

**Comment by stgeke** [comment]
How about `rho` and `rhoCp`?

**Reply by bnkeeton** [reply]
They are both set to a constant non-dimensional value of 1 in `fillProp` and `uservp`.

**Comment by bnkeeton** [comment]
Hi Stefan, I'm curious if there is any update? I tried using the next branch v24.0.2 (sha:0003f4ae) and also find a solution that differs considerably from nek5000.

The case files originally provided are set up to run the single time step and output the divergence. I'm also happy to run any additional tests that might help resolve the discrepancy.

Thanks,
Ben

**Reply by stgeke** [reply]
Can you please try the very latest `next`

**Reply by bnkeeton** [reply]
Yes, I just tested the latest `next` branch: v24.0.2 (sha:c6a366da). The divergence is now fixed, and it is identical to that given by nek5000. I will continue to run more complex test cases, and will report back if I notice any strange behavior.

Thanks,
Ben

**Comment by raj-brown** [comment]
Hi @stgeke @bnkeeton I started recently a project for low mach and I was wondering If I can get some reference paper containing the system of equations solved in NekRS low mach solver. Thank you very much!

**Comment by fischer1** [comment]
There is a chapter here:
[cid:8ba6557d-4b34-4a77-bed1-34cf9779b08d]

And another 2017 article here:

https://www.mcs.anl.gov/~fischer/pubhtml/ccfd.pdf


Recent Developments in Spectral Element Simulations of Moving-Domain Problems - Argonne National Laboratory<https://www.mcs.anl.gov/~fischer/pubhtml/ccfd.pdf>
Recent Developments in Spectral Element Simulations of Moving-Domain Problems Paul Fischer, Martin Schmitt, and Ananias Tomboulides Abstract Presented here are recent developments in spectral element methods for
www.mcs.anl.gov
hth,
Paul

________________________________

**Comment by raj-brown** [comment]
Thanks Prof. Fischer @fischer1 !  I think the link to chapter got mangled It shows ```id:8ba6557d-4b34-4a77-bed1-34cf9779b08d]```. Could you please repost? Thanks!

**Reply by yslan** [reply]
Nek5000 doc has the equations. 
https://nek5000.github.io/NekDoc/theory.html#low-mach-navier-stokes

Chap 3.5 has low-mach solved by PN-PN scheme.
https://doi.org/10.1016/B978-0-32-391144-3.00009-7

**Reply by raj-brown** [reply]
thanks @yslan. I just had a code design question on plugin for ```lowmach.cpp```. How the plugin is interacting with main ns solver? Thank you!

**Reply by yslan** [reply]
For the current repo/next branch (v24pre), see the example `lowMach/lowMach.udf`.
lowMach add the thermal material derivative into the divergence `nrs->userDivergence`. It's updated via the call `this->evaluateDivergence(timeNew);` in `nrs/timeStepper.cpp`.

**Reply by raj-brown** [reply]
Thanks @yslan. I appreciate it.

**Reply by raj-brown** [reply]
Hi @yslan I had a question as I am going through the codes of nekRS. What is the roled [nekInterface](https://github.com/Nek5000/nekRS/tree/next/src/nekInterface) and [pointInterpolation](https://github.com/Nek5000/nekRS/tree/next/src/pointInterpolation). Thanks!

**Reply by yslan** [reply]
nekInterface is mostly the Fortran interface that calls Nek5000's subroutines. It (`nek::`) contains reader/writer, user subroutines like `usrdat2` and `userchk` and the initial setup routine for mesh, connectivity, boundary conditions.

pointInterpolation is for interpolating the field value onto arbitrary points.
With defined points, it locates the MPI rank owning the elements containing a point, solve the reference coordinates, and interpolate from SEM mesh onto those points. 
See this [example](https://github.com/Nek5000/nekRS/blob/5152fec972a059936a5c7b68c3b4b2772248bf45/examples/ktauChannel/channel.udf#L13) that sampling the values on a line.

**Reply by raj-brown** [reply]
Thanks @yslan Really appreciate! Another question I had on Example ```mv_cyl```. Could you please point me to a paper/reference containing the problem details.Thank you very much!

**Reply by yslan** [reply]
I believe it was a Nek5000 example from [here](https://github.com/Nek5000/NekExamples/tree/master/mv_cyl). However, I don't know where the case coming from. You might want to ask in Nek5000's [Google group](https://groups.google.com/g/nek5000).
